File: heatmap_stats.rb
require 'json'

def stats(vals)
  {count: vals.size, mean: vals.sum(0.0) / vals.size }  #, values: vals}
end

data = JSON.parse(File.read('heatmaps/CTCF.json'))['data']
triples = data.flat_map{|ds_info|
  ds_1 = ds_info['name']
  ds_info['data'].map{|corr_info|
    ds_2 = corr_info['x']
    val = corr_info['y']
    [ds_1, ds_2, val]
  }
}.select{|ds_1, ds_2, val|
  ds_1 < ds_2
}.map{|ds_1, ds_2, vals|
  [
    [ds_1.split(':').first, ds_1],
    [ds_2.split(':').first, ds_2],
    vals,
  ]
}

data_types = triples.flat_map{|(dt_1, ds_1), (dt_2, ds_2), val|
  [dt_1, dt_2]
}.uniq

data_types.repeated_combination(2).map{|data_type_1_2|
  matching_triples = triples.select{|(dt_1, ds_1), (dt_2, ds_2), val|
    [dt_1, dt_2].sort == data_type_1_2.sort
  }
  vals = matching_triples.map(&:last)
  [*data_type_1_2.sort, stats(vals)]
}

File: heatmap_stats.py
import json
import itertools

def mean(vals):
    if len(vals) == 0:
        return None
    return sum(vals) / len(vals)

def stats(vals):
    return {'count': len(vals), 'mean': mean(vals), 'values': vals}

with open('heatmaps/CTCF.json') as f:
    data = json.load(f)['data']

triples = []
for ds_info in data:
    ds_1 = ds_info['name']
    for corr_info in ds_info['data']:
        ds_2 = corr_info['x']
        val = corr_info['y']
        if ds_1 < ds_2:
            triples.append(((ds_1.split(':')[0], ds_1), (ds_2.split(':')[0], ds_2), val))

data_types = set()
for ((dt_1, ds_1), (dt_2, ds_2), val) in triples:
    data_types.add(dt_1)
    data_types.add(dt_2)

dt_stats = {}
for data_types_pair in itertools.combinations_with_replacement(data_types, 2):
    # matching_triples = [((dt_1, ds_1), (dt_2, ds_2), val)  for ((dt_1, ds_1), (dt_2, ds_2), val) in triples  if sorted((dt_1, dt_2)) == sorted(data_types_pair)]
    vals = [val  for ((dt_1, ds_1), (dt_2, ds_2), val) in triples  if sorted((dt_1, dt_2)) == sorted(data_types_pair)]
    info = [*sorted(data_types_pair), stats(vals)]
    dt_stats[tuple(sorted(data_types_pair))] = stats(vals)
    print(info)
